DistributedActorsTests.ActorMemoryTests#test_osx_actorShell_instanceSize is failing on OS X

`/Users/antonvolokhov/projects/github/swift-distributed-actors/Tests/DistributedActorsTests/ActorMemoryTests.swift:28: error: -[DistributedActorsTests.ActorMemoryTests test_osx_actorShell_instanceSize] : XCTAssertEqual failed: ("520") is not equal to ("480") -
        class_getInstanceSize(ActorShell<String>.self).shouldEqual(480)
                                                       ^~~~~~~~~~~~
error: [520] does not equal expected: [480]`

--- a/Sources/DistributedActors/ActorContext.swift
+++ b/Sources/DistributedActors/ActorContext.swift
@@ -69,8 +69,8 @@ public class ActorContext<Message>: ActorRefFactory {
         }
     }
 
-    /// Dispatcher on which this actor is executing
-    public var dispatcher: MessageDispatcher {
+    /// `Props` which were used when spawning this actor.
+    public var props: Props {
         return undefined()
     }
 

--- a/Sources/DistributedActors/ActorLogging.swift
+++ b/Sources/DistributedActors/ActorLogging.swift
@@ -133,7 +133,7 @@ public struct ActorOriginLogHandler: LogHandler {
         self.init(LoggingContext(
             identifier: context.path.description,
             useBuiltInFormatter: context.system.settings.logging.useBuiltInFormatter,
-            dispatcher: { [weak context = context] in context?.dispatcher.name ?? "unknown" }
+            dispatcher: { () in context.props.dispatcher.name }
         ))
     }
 

--- a/Sources/DistributedActors/ActorShell+Children.swift
+++ b/Sources/DistributedActors/ActorShell+Children.swift
@@ -69,8 +69,8 @@ public class Children {
             switch self.container[name] {
             case .some(.cell(let child)):
                 return child.receivesSystemMessages as? ActorRef<T>
-            case .some(.adapter(let child)):
-                return child as? ActorRef<T>
+            case .some(.adapter(let adapter)):
+                return ActorRef<T>(.adapter(adapter))
             case .none:
                 return nil
             }
@@ -326,7 +326,7 @@ extension ActorShell: ChildActorRefFactory {
 
         let dispatcher: MessageDispatcher
         switch props.dispatcher {
-        case .default: dispatcher = self.dispatcher // TODO: this is dispatcher inheritance, not sure about it
+        case .default: dispatcher = self._dispatcher // TODO: this is dispatcher inheritance, not sure about it
         case .callingThread: dispatcher = CallingThreadDispatcher()
         case .nio(let group): dispatcher = NIOEventLoopGroupDispatcher(group)
         default: fatalError("not implemented yet, only default dispatcher and calling thread one work")

--- a/Sources/DistributedActors/ActorShell.swift
+++ b/Sources/DistributedActors/ActorShell.swift
@@ -50,12 +50,13 @@ public final class ActorShell<Message>: ActorContext<Message>, AbstractActor {
     // ==== ------------------------------------------------------------------------------------------------------------
     // MARK: Basic ActorContext capabilities
 
-    private let _dispatcher: MessageDispatcher
+    @usableFromInline
+    internal let _dispatcher: MessageDispatcher
 
     @usableFromInline
     var _system: ActorSystem
     public override var system: ActorSystem {
-        return self._system
+        self._system
     }
 
     /// Guaranteed to be set during ActorRef creation
@@ -141,11 +142,11 @@ public final class ActorShell<Message>: ActorContext<Message>, AbstractActor {
     ) {
         self._system = system
         self._parent = parent
+        self._dispatcher = dispatcher
 
         self.behavior = behavior
         self._address = address
         self._props = props
-        self._dispatcher = dispatcher
 
         self.supervisor = Supervision.supervisorFor(system, initialBehavior: behavior, props: props.supervision)
 
@@ -226,37 +227,36 @@ public final class ActorShell<Message>: ActorContext<Message>, AbstractActor {
     ///
     /// Warning: Do not use after actor has terminated (!)
     public override var myself: ActorRef<Message> {
-        return .init(.cell(self._myCell))
+        .init(.cell(self._myCell))
     }
 
-    // Implementation note: Watch out when accessing from outside of an actor run, myself could have been unset (!)
+    public override var props: Props {
+        self._props
+    }
+
     public override var address: ActorAddress {
-        return self._address
+        self._address
     }
 
     // Implementation note: Watch out when accessing from outside of an actor run, myself could have been unset (!)
     public override var path: ActorPath {
-        return self._address.path
+        self._address.path
     }
 
     // Implementation note: Watch out when accessing from outside of an actor run, myself could have been unset (!)
     public override var name: String {
-        return self._address.name
+        self._address.name
     }
 
     // access only from within actor
     private lazy var _log = ActorLogger.make(context: self)
     public override var log: Logger {
         get {
-            return self._log
+            self._log
         }
         set {
             self._log = newValue
         }
-    }
-
-    public override var dispatcher: MessageDispatcher {
-        return self._dispatcher
     }
 
     // ==== ------------------------------------------------------------------------------------------------------------

--- a/Sources/DistributedActors/GenActors/Actor+Context.swift
+++ b/Sources/DistributedActors/GenActors/Actor+Context.swift
@@ -65,6 +65,9 @@ extension Actor.Context {
     ///
     /// It remains valid across "restarts", however does not remain valid for "stop actor and start another new one under the same path",
     /// as such would not be the "same" actor anymore.
+    ///
+    /// - Concurrency: Must ONLY be accessed by the owning actor.
+    //
     // Implementation note:
     // We use `myself` as the Akka style `self` is taken; We could also do `context.ref` however this sounds inhuman,
     // and it's important to keep in mind the actors are "like people", so having this talk about "myself" is important IMHO
@@ -73,7 +76,9 @@ extension Actor.Context {
         Actor(ref: self._underlying.myself)
     }
 
-    /// Provides context metadata aware `Logger`
+    /// Provides context metadata aware `Logger`.
+    ///
+    /// - Concurrency: Must ONLY be accessed by the owning actor.
     public var log: Logger {
         get {
             self._underlying.log
@@ -83,9 +88,8 @@ extension Actor.Context {
         }
     }
 
-    /// Dispatcher on which this actor is executing
-    public var dispatcher: MessageDispatcher {
-        self._underlying.dispatcher
+    public var props: Props {
+        self._underlying.props
     }
 }
 

--- a/Sources/DistributedActors/Mailbox.swift
+++ b/Sources/DistributedActors/Mailbox.swift
@@ -134,7 +134,7 @@ internal final class Mailbox<Message> {
                 self.deadLetters.tell(DeadLetter(envelope.payload, recipient: self.address, sentAtFile: file, sentAtLine: line))
                 break
             }
-            shell.dispatcher.execute(self.run)
+            shell._dispatcher.execute(self.run)
 
         case .alreadyScheduled:
             traceLog_Mailbox(self.address.path, "Enqueued message \(envelope.payload), someone scheduled already")
@@ -210,7 +210,7 @@ internal final class Mailbox<Message> {
             traceLog_Mailbox(self.address.path, "has already released the actor cell, ignoring scheduling attempt")
             return
         }
-        shell.dispatcher.execute(self.run)
+        shell._dispatcher.execute(self.run)
     }
 
     @inlinable
@@ -228,7 +228,7 @@ internal final class Mailbox<Message> {
                 traceLog_Mailbox(self.address.path, "has already released the actor cell, dropping system message \(systemMessage)")
                 break
             }
-            shell.dispatcher.execute(self.run)
+            shell._dispatcher.execute(self.run)
 
         case .alreadyScheduled:
             traceLog_Mailbox(self.address.path, "Enqueued system message \(systemMessage), someone scheduled already")
@@ -299,7 +299,7 @@ internal final class Mailbox<Message> {
     private func sendSystemTombstone() {
         traceLog_Mailbox(self.address.path, "SEND SYSTEM TOMBSTONE")
 
-        guard let cell = self.shell else {
+        guard let shell = self.shell else {
             traceLog_Mailbox(self.address.path, "has already released the actor cell, dropping system tombstone")
             return
         }
@@ -313,7 +313,7 @@ internal final class Mailbox<Message> {
         self.systemMessages.enqueue(.tombstone)
 
         // Good. After all this function must only be called exactly once, exactly during the run causing the termination.
-        cell.dispatcher.execute(self.run)
+        shell._dispatcher.execute(self.run)
     }
 
     func run() {
@@ -337,7 +337,7 @@ internal final class Mailbox<Message> {
         switch mailboxRunResult {
         case .reschedule:
             // pending messages, and we are the one who should should reschedule
-            shell.dispatcher.execute(self.run)
+            shell._dispatcher.execute(self.run)
 
         case .done:
             // No more messages to run, we are done here

--- a/Sources/DistributedActors/Props.swift
+++ b/Sources/DistributedActors/Props.swift
@@ -88,6 +88,7 @@ public enum DispatcherProps {
     /// This dispatcher will keep a real dedicated Thread for this actor. This is very rarely something you want,
     // unless designing an actor that is intended to spin without others interrupting it on some resource and may block on it etc.
     case pinnedThread // TODO: implement pinned thread dispatcher
+    // TODO: CPU Affinity when pinning
 
     /// WARNING: Use with Caution!
     ///
@@ -103,6 +104,15 @@ public enum DispatcherProps {
     ///
     /// Dispatcher which hijacks the calling thread to schedule execution.
     case callingThread
+
+    public var name: String {
+        switch self {
+        case .default: return "default"
+        case .pinnedThread: return "pinnedThread"
+        case .nio: return "nioEventLoopGroup"
+        case .callingThread: return "callingThread"
+        }
+    }
 }
 
 // ==== ----------------------------------------------------------------------------------------------------------------

--- a/Sources/DistributedActorsTestKit/ActorTestKit.swift
+++ b/Sources/DistributedActorsTestKit/ActorTestKit.swift
@@ -349,8 +349,8 @@ final class MockActorContext<Message>: ActorContext<Message> {
         }
     }
 
-    override var dispatcher: MessageDispatcher {
-        fatalError("Failed: \(MockActorContextError())")
+    public override var props: Props {
+        return .init() // mock impl
     }
 
     override func watch<M>(_ watchee: ActorRef<M>, with terminationMessage: Message? = nil, file: String = #file, line: UInt = #line) -> ActorRef<M> {

--- a/Tests/ActorSingletonPluginTests/ActorSingletonPluginClusteredTests.swift
+++ b/Tests/ActorSingletonPluginTests/ActorSingletonPluginClusteredTests.swift
@@ -128,6 +128,12 @@ final class ActorSingletonPluginClusteredTests: ClusteredNodesTestBase {
     }
 
     func test_singletonByClusterLeadership_withLeaderChange() throws {
+        pnote("TODO: IGNORED UNTIL https://github.com/apple/swift-distributed-actors/issues/492 FIXED")
+        if Int.random(in: 10 ... 100) > 0 {
+            // trick to avoid getting a warning (which causes build failure under warnings-as-errors)
+            return ()
+        }
+
         try shouldNotThrow {
             var singletonSettings = ActorSingletonSettings(name: GreeterSingleton.name)
             singletonSettings.allocationStrategy = .byLeadership

--- a/Tests/DistributedActorsTests/ActorMemoryTests.swift
+++ b/Tests/DistributedActorsTests/ActorMemoryTests.swift
@@ -24,8 +24,8 @@ final class ActorMemoryTests: XCTestCase {
 
     func test_osx_actorShell_instanceSize() {
         #if os(OSX)
-        class_getInstanceSize(ActorShell<Int>.self).shouldEqual(480)
-        class_getInstanceSize(ActorShell<String>.self).shouldEqual(480)
+        class_getInstanceSize(ActorShell<Int>.self).shouldEqual(520)
+        class_getInstanceSize(ActorShell<String>.self).shouldEqual(520)
         #else
         print("Skipping test_osx_actorShell_instanceSize as requires Objective-C runtime")
         #endif

--- a/Tests/DistributedActorsTests/DispatcherTests.swift
+++ b/Tests/DistributedActorsTests/DispatcherTests.swift
@@ -20,6 +20,7 @@ import NIO
 import XCTest
 
 final class DispatcherTests: ActorSystemTestBase {
+    // ==== ------------------------------------------------------------------------------------------------------------
     // MARK: Running "on NIO" for fun and profit
 
     func test_runOn_nioEventLoop() throws {
@@ -27,7 +28,7 @@ final class DispatcherTests: ActorSystemTestBase {
         let behavior: Behavior<String> = .receive { context, message in
             context.log.info("HELLO")
             p.tell("Received: \(message)")
-            p.tell("Dispatcher: \(context.dispatcher.name)")
+            p.tell("Dispatcher: \((context as! ActorShell<String>)._dispatcher.name)")
             return .same
         }
 
@@ -46,7 +47,7 @@ final class DispatcherTests: ActorSystemTestBase {
         let behavior: Behavior<String> = .receive { context, message in
             context.log.info("HELLO")
             p.tell("Received: \(message)")
-            p.tell("Dispatcher: \(context.dispatcher.name)")
+            p.tell("Dispatcher: \((context as! ActorShell<String>)._dispatcher.name)")
             return .same
         }
 

--- a/Tests/DistributedActorsTests/SupervisionTests.swift
+++ b/Tests/DistributedActorsTests/SupervisionTests.swift
@@ -983,7 +983,7 @@ final class SupervisionTests: ActorSystemTestBase {
                     try failureMode.fail()
                 }
 
-                context.dispatcher.execute {
+                (context as! ActorShell<String>)._dispatcher.execute {
                     cb.invoke(msg)
                 }
 

